src/compositemark/boxplot.ts

```diff
@@ -165,72 +153,71 @@ export function normalizeBoxPlot(spec: GenericUnitSpec<Encoding<Field>, BOXPLOT
-    resultOrient = 'vertical';
-  } else if (isDiscrete(yDef) && isContinuous(xDef)) {
-    resultOrient = 'horizontal';
-  } else {
-    if (isContinuous(xDef) && isContinuous(yDef)) {
-      if (xDef.aggregate === undefined && yDef.aggregate === BOXPLOT) {
-        resultOrient = 'vertical';
-      } else if (yDef.aggregate === undefined && xDef.aggregate === BOXPLOT) {
-        resultOrient = 'horizontal';
-      } else if (xDef.aggregate === BOXPLOT && yDef.aggregate === BOXPLOT) {
-        throw new Error('Both x and y cannot have aggregate');
-      } else {
-        if (isBoxPlotDef(mark)) {
-          if (mark && mark.orient) {
-            resultOrient = mark.orient;
-          } else {
-            // default orientation = vertical
-            resultOrient = 'vertical';
-          }
-        } else {
-          resultOrient = 'vertical';
-        }
-      }
+  if (continuousXAxis
+      && (encoding.y === undefined || (isFieldDef(encoding.y) && isDiscrete(encoding.y)))) {
+    return 'horizontal';
+  } else if (continuousYAxis
+      && (encoding.x === undefined || (isFieldDef(encoding.x) && isDiscrete(encoding.x)))) {
+    return 'vertical';
+  }
+
+  if (!continuousXAxis && !continuousYAxis) {
+    throw new Error('Need a valid continuous axis for boxplots');
+  }
+
+  if (isFieldDef(encoding.x) && isFieldDef(encoding.y) && isContinuous(encoding.x) && isContinuous(encoding.y)) {
+    if (encoding.x.aggregate === undefined && encoding.y.aggregate === BOXPLOT) {
+      return 'vertical';
+    } else if (encoding.y.aggregate === undefined && encoding.x.aggregate === BOXPLOT) {
+      return 'horizontal';
+    } else if (encoding.x.aggregate === BOXPLOT && encoding.y.aggregate === BOXPLOT) {
+      throw new Error('Both x and y cannot have aggregate');
     } else {
-      throw new Error('Both x and y cannot be discrete');
+      if (isBoxPlotDef(mark) && mark.orient) {
+        return mark.orient;
+      }
     }
   }
 
-  return resultOrient;
+  // default orientation = vertical
+  return 'vertical';
 }
 
 
-export function box2DParams(spec: GenericUnitSpec<Encoding<Field>, BOXPLOT | BoxPlotDef>, orient: Orient) {
+export function boxParams(spec: GenericUnitSpec<Encoding<Field>, BOXPLOT | BoxPlotDef>, orient: Orient) {
   const {mark: mark, encoding: encoding, ...outerSpec} = spec;
 
-  let discreteAxisFieldDef: PositionFieldDef<Field>;
+  let discreteAxisChannelDef: PositionFieldDef<Field>;
   let continuousAxisChannelDef: PositionFieldDef<Field>;
   let discreteAxis;
   let continuousAxis;
 
-  // FIXME: refactor code such that we don't have to do this casting
-  // We can cast here as we already check from outside that both x and y are FieldDef
-  const xDef = encoding.x as FieldDef<Field>;
-  const yDef = encoding.y as FieldDef<Field>;
-
-
   if (orient === 'vertical') {
-    discreteAxis = 'x';
-    continuousAxis = 'y';
-    continuousAxisChannelDef = yDef;
-    discreteAxisFieldDef = xDef;
+    if (isFieldDef(encoding.y)) {
```

Comment:
@kanitw yes I have that only because I need to typeguard encoding.y to ensure it is of type FieldDef<Field> which I assign to continuousAxisChannelDef